packages/squiggle-lang/src/fr/danger.ts

```diff
@@ -33,6 +34,27 @@ const maker = new FnFactory({
   requiresNamespace: true,
 });
 
+function combinations<T>(arr: T[], k: number): T[][] {
+  if (k === 0) return [[]];
+  if (k === arr.length) return [arr];
+
+  const withoutFirst = combinations(arr.slice(1), k);
+  const withFirst = combinations(arr.slice(1), k - 1).map((comb) => [
+    arr[0],
+    ...comb,
+  ]);
+
+  return withFirst.concat(withoutFirst);
+}
+
+function allCombinations<T>(arr: T[]): T[][] {
+  let allCombs: T[][] = [];
+  for (let k = 1; k <= arr.length; k++) {
+    allCombs = allCombs.concat(combinations(arr, k));
+  }
+  return allCombs;
+}
```

Comment:
This doesn't matter much and your version seems correct to me, but here's alternative way I'd implement this:

```
function allCombinations(arr) {
  const result = [];
  for (let i = 1; i < 1 << arr.length; i++) {
    const cur = [];
    for (let p = 0; p < arr.length; p++) {
      if (i & (1 << p)) {
        cur.push(arr[p]);
      }
    }
    result.push(cur);
  }
  return result;
}
```

It might look scary because of bit arithmetics, but it's simpler conceptually once you realize that all you need is to iterate through all possible bitmasks (there are 2^N of those, except the empty one, so from `1` up to `1 << arr.length`) and then for each mask take the element iff there's `1` in that position.

Reply:
Another note: my version won't work correctly for `arr.length > 30`, but you might want to throw on those arrays anyway; combinations of 30 elements would take >250GB of memory to store.

Reply:
Huh, that's neat. I actually got my code from ChatGPT. Happy to try yours.

Reply:
One more thing: maybe this function should generate an empty combination too? I believe the standard definition includes it (`C(n, 0)` is 1, not 0).

Reply:
I tried your algorithm. It returns the element in a different order. 
```
Danger.allCombinations([3,5,8])
```

That produced:
``[[3],[5],[8],[3,5],[3,8],[5,8],[3,5,8]]`` 

Instead of 
``[[3],[5],[3,5],[8],[3,8],[5,8],[3,5,8]]``

I think I prefer the former result, so am keeping for now. I guess we could also just sort it after, but that could produce other complications.

Feel free to find some way to change, then submit a change after this. 

I think we shouldn't spend much time on this now though.

Reply:
I did a patched verison locally:

```
function allCombinations<T>(arr: T[]): T[][] {
  const resultBySize: T[][][] = new Array<T[][]>(arr.length + 1)
    .fill([]) // for type safety, will be overwritten by map
    .map(() => []);

  for (let i = 1; i < 1 << arr.length; i++) {
    const cur = [];
    for (let p = 0; p < arr.length; p++) {
      if (i & (1 << p)) {
        cur.push(arr[p]);
      }
    }
    resultBySize[cur.length].push(cur);
  }
  return ([] as T[][]).concat(...resultBySize);
}
```

But then I noticed that you also added `Danger.combinations`, and code reuse won't be as nice (I mean, I could return `resultBySize[n]`, but that would be computationally sub-optimal and sometimes terrible, it's useful to ask for combinations of size 2 for a long array, and `C(10, 2)` is much less than `10!`), so let's skip it for now.